### PR TITLE
Remove docs for ActorPub/Sub and write persistence query docs as stages

### PR DESCRIPTION
**Pull request by chbatey**
Main change is to re-write the events by tag examples as stages 

Refs #26187

### CODE REVIEW

File: akka-docs/src/main/paradox/persistence-query.md [hunk 1]
```diff
@@ -303,10 +303,10 @@ Java
 And the `eventsByTag` could be backed by such an Actor for example:
 
 Scala
-:  @@snip [MyEventsByTagPublisher.scala](/akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagPublisher.scala) { #events-by-tag-publisher }
+:  @@snip [MyEventsByTagSource.scala](/akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala) { #events-by-tag-publisher }
```
Comment (johanandren) [inline akka-docs/src/main/paradox/persistence-query.md]: So now it is not "such an Actor" any more

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      this.initialOffset = initialOffset;
+      this.refreshInterval = refreshInterval;
+   }
+
+   @Override
+   public SourceShape<EventEnvelope> shape() {
+      return SourceShape.of(out);
+   }
+
+   class Logic extends TimerGraphStageLogic implements OutHandler {
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: package private? Should it be public or private rather? Or just anonymous instance like we do with the scala code?

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+      private ActorSystem system = ((ActorMaterializer) materializer()).system();
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Not nice with the cast in example code, but `akka.stream.ActorMaterializerHelper#downcast` is internal...

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+           push(out, buf.remove(0));
+         }
+      }
+
+      private void query() {
+         if (buf.isEmpty()) {
+            final String query =
+               "SELECT id, persistent_repr "
+                  + "FROM journal WHERE tag = ? AND id > ? "
+                  + "ORDER BY id LIMIT ?";
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Never changes so could be made a constant

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+// #events-by-tag-publisher
+public class MyEventsByTagSource extends GraphStage<SourceShape<EventEnvelope>> {
+   private Outlet<EventEnvelope> out = Outlet.create("MyEventByTagSource.out");
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: shouldn't this be public to allow usage of the source in the GraphDSL?

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+   private static final String CONTINUE = "CONTINUE";
+   private static final int LIMIT = 1000;
+   private final Connection connection;
```
Comment (johanandren) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Should the connection be shared by each materialized source or should there be a separate connection?

File: akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      override def onDownstreamFinish(): Unit = {
+        // close connection if responsible for doing so
+      }
+
+      private def query(): Unit = {
+        if (buf.isEmpty) {
+          try {
+            val result = Select.run(tag, currentOffset, Limit)
+            currentOffset = if (result.nonEmpty) result.last._1 else currentOffset
+            val serialization = SerializationExtension(system)
```
Comment (johanandren) [inline akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala]: Look it up once on mat like the Java one instead of every query?

**Review by johanandren — COMMENTED** [review]
Look'n good. Nitpicking.

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+// #events-by-tag-publisher
+public class MyEventsByTagSource extends GraphStage<SourceShape<EventEnvelope>> {
+   private Outlet<EventEnvelope> out = Outlet.create("MyEventByTagSource.out");
+   private static final String CONTINUE = "CONTINUE";
```
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: use  enum
```
static enum Continue {
  INSTANCE;
}

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+      private void deliver() {
+         if (isAvailable(out) && !buf.isEmpty()) {
+           push(out, buf.remove(0));
+         }
+      }
+
+      private void query() {
+         if (buf.isEmpty()) {
+            final String query =
+               "SELECT id, persistent_repr "
```
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: we should change this example have have one column for event payload (serialized bytes) and separate columns for other meta data, such as `persistence_id`, `seq_nr`, `serializer_id`, `serializer_manifest`

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+                  buf =
+                     res.stream()
+                        .map(
+                           in -> {
+                              final Long id = in.first();
+                              final byte[] bytes = in.second();
+
+                              final PersistentRepr p =
+                                 serialization.deserialize(bytes, PersistentRepr.class).get();
```
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: see comment above regarding separate columns

File: akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case Continue =>
+          query()
+          tryPush()
+      }
+
+      object Select {
+        private def statement() = connection.prepareStatement("""
+        SELECT id, persistent_repr FROM journal
```
Comment (patriknw) [inline akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala]: same here ofc

File: akka-docs/src/main/paradox/persistence-query.md (hunk 1 above)
Comment (chbatey) [reply akka-docs/src/main/paradox/persistence-query.md]: Indeed!

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java (hunk 2 above)
Comment (chbatey) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: I got hung up on an anonymous class can't also implement an interface but i'll change the OutHander to be a separate anonymous class

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java (hunk 3 above)
Comment (chbatey) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Indeed

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java (hunk 6 above)
Comment (chbatey) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: I'll put a comment highlighting this, I just followed the previous example

File: akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala (hunk 7 above)
Comment (chbatey) [reply akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala]: Doh, it wasn't me

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java (hunk 9 above)
Comment (chbatey) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Yep good idea. I was blindly copying from the old example

File: akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java (hunk 2 above)
Comment (johanandren) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Isn't there just some funky syntax for that?

Comment (johanandren) [reply akka-docs/src/test/java/jdocs/persistence/query/MyEventsByTagSource.java]: Oh:
> An anonymous inner class can extend one subclass or implement one interface

No funky syntax then

File: akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+import scala.util.control.NonFatal
+
+//#events-by-tag-publisher
+class MyEventsByTagSource(tag: String, offset: Long, refreshInterval: FiniteDuration)
+    extends GraphStage[SourceShape[EventEnvelope]] {
+
+  private case object Continue
+  val out: Outlet[EventEnvelope] = Outlet("MyEventByTagSource.out")
+  override def shape: SourceShape[EventEnvelope] = SourceShape(out)
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
```
Comment (patriknw) [inline akka-docs/src/test/scala/docs/persistence/query/MyEventsByTagSource.scala]: since this is blocking it should use blockingIoDispatcher

**Review by patriknw — APPROVED** [review]
LGTM, thanks for the additional improvements

**Comment by patriknw** [comment]
@chbatey something messed up with mima filters
```
[error] Error while parsing /home/travis/build/akka/akka/akka-stream/src/main/mima-filters/2.5.x.backwards.excludes, line 64: Couldn't parse ' dispatcher aliases made internal streams dispatcher resolve superfluous #26775'
```

**Comment by chbatey** [comment]
Messed up the merge, fixing

**Comment by patriknw** [comment]
🎆

**Pull request merged**